File: old-julia/src/lattice.jl
# src/lattice.jl
#
# Lattice mathematics: strut directions, node/basis geometry, diamond profile
# frames, and candidate-strut enumeration from a world-space AABB. Implements
# docs/algorithm.md §2 and §3.1 exactly — do not approximate these formulas.

using LinearAlgebra

# --- Lightweight 3D vector type --------------------------------------------
#
# A small immutable, isbits struct (not a type-pirated NTuple/Base operator
# overload) used throughout lattice.jl, classify.jl, tiling.jl, and
# geomkernel.jl for cheap, allocation-free 3D vector arithmetic.

struct Vec3
    x::Float64
    y::Float64
    z::Float64
end

Vec3(t::NTuple{3,<:Real}) = Vec3(Float64(t[1]), Float64(t[2]), Float64(t[3]))

Base.:+(a::Vec3, b::Vec3) = Vec3(a.x + b.x, a.y + b.y, a.z + b.z)
Base.:-(a::Vec3, b::Vec3) = Vec3(a.x - b.x, a.y - b.y, a.z - b.z)
Base.:-(a::Vec3) = Vec3(-a.x, -a.y, -a.z)
Base.:*(s::Real, a::Vec3) = Vec3(s * a.x, s * a.y, s * a.z)
Base.:*(a::Vec3, s::Real) = s * a
Base.:/(a::Vec3, s::Real) = Vec3(a.x / s, a.y / s, a.z / s)
Base.isapprox(a::Vec3, b::Vec3; kwargs...) =
    isapprox(a.x, b.x; kwargs...) && isapprox(a.y, b.y; kwargs...) && isapprox(a.z, b.z; kwargs...)
Base.Tuple(a::Vec3) = (a.x, a.y, a.z)

dot3(a::Vec3, b::Vec3) = a.x * b.x + a.y * b.y + a.z * b.z
cross3(a::Vec3, b::Vec3) = Vec3(a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x)
norm3(a::Vec3) = sqrt(dot3(a, a))
normalize3(a::Vec3) = a / norm3(a)

# --- Strut recline angle (docs/algorithm.md §2.1) --------------------------

"""Exact strut recline angle from +Z, computed natively — not a decimal literal."""
const THETA = asin(sqrt(2 / 3))
const COS_THETA = inv(sqrt(3))
const SIN_THETA = sqrt(2 / 3)

"""
    strut_direction(k::Integer) -> Vec3

Unit direction vector for strut family `k ∈ {0,1,2}`, at azimuth `2πk/3`,
reclined from +Z by `THETA` (docs/algorithm.md §2.1).
"""
function strut_direction(k::Integer)
    φ = 2π * k / 3
    return Vec3(SIN_THETA * cos(φ), SIN_THETA * sin(φ), COS_THETA)
end

"""Cube edge length `a = cc / √2` (docs/algorithm.md §2.2)."""
cube_edge(cc::Real) = cc / sqrt(2)

# --- Precomputed lattice geometry ------------------------------------------

"""
    LatticeParams(cc, t)

All lattice geometry derived from the CLI parameters `cc` and `t`
(docs/algorithm.md §2, §3.1). Direction/profile-axis tuples are indexed
`1:3`, corresponding to strut families `e0, e1, e2` (`k = 0, 1, 2`) in the
math notation.
"""
struct LatticeParams
    cc::Float64
    t::Float64
    a::Float64                    # cube edge length
    directions::NTuple{3,Vec3}    # e0, e1, e2 (unit vectors)
    u::NTuple{3,Vec3}             # horizontal profile axis per direction
    v::NTuple{3,Vec3}             # vertical-plane profile axis per direction
    r::Float64                    # strut circumradius = t/√2
    B::NTuple{3,Vec3}             # basis columns: node(i,j,k) = i*B[1]+j*B[2]+k*B[3]
end

function LatticeParams(cc::Real, t::Real)
    a = cube_edge(cc)
    dirs = ntuple(m -> strut_direction(m - 1), 3)
    zhat = Vec3(0.0, 0.0, 1.0)
    us = ntuple(m -> normalize3(cross3(zhat, dirs[m])), 3)
    vs = ntuple(m -> cross3(dirs[m], us[m]), 3)
    r = t / sqrt(2)
    B = ntuple(m -> a * dirs[m], 3)
    return LatticeParams(Float64(cc), Float64(t), a, dirs, us, vs, r, B)
end

"""
    node(lp, i, j, k) -> Vec3

World-space position of lattice node `(i,j,k)` (docs/algorithm.md §2.2).
"""
node(lp::LatticeParams, i::Integer, j::Integer, k::Integer) =
    i * lp.B[1] + j * lp.B[2] + k * lp.B[3]

"""
    profile_vertices(lp, center, dir) -> NTuple{4,Vec3}

The four corner points of the diamond cross-section profile centered at
`center`, for strut direction index `dir ∈ 1:3` (docs/algorithm.md §3.1).
"""
function profile_vertices(lp::LatticeParams, center::Vec3, dir::Integer)
    hu = lp.r * lp.u[dir]
    hv = lp.r * lp.v[dir]
    return (center + hu, center + hv, center - hu, center - hv)
end

# --- Strut references and enumeration --------------------------------------

"""A strut identified by its origin node index and direction (`1:3`)."""
struct StrutRef
    i::Int
    j::Int
    k::Int
    dir::Int
end

"""World-space endpoints `(p0, p1)` of a strut, `p1 - p0 == lp.B[s.dir]`."""
function endpoints(lp::LatticeParams, s::StrutRef)
    p0 = node(lp, s.i, s.j, s.k)
    p1 = p0 + lp.B[s.dir]
    return p0, p1
end

"""World-space midpoint of a strut segment."""
function midpoint(lp::LatticeParams, s::StrutRef)
    p0, p1 = endpoints(lp, s)
    return 0.5 * (p0 + p1)
end

"""
    index_range(lp, lo, hi) -> (lo_idx, hi_idx)

Inclusive integer index range, as `NTuple{3,Int}` bounds, covering every
lattice node whose emanating struts could intersect the world-space AABB
`[lo, hi]`, padded by the strut circumradius (docs/algorithm.md §2.4).
"""
function index_range(lp::LatticeParams, lo::Vec3, hi::Vec3)
    Bmat = hcat(collect(Tuple(lp.B[1])), collect(Tuple(lp.B[2])), collect(Tuple(lp.B[3])))
    corners = (Vec3(x, y, z) for x in (lo.x, hi.x), y in (lo.y, hi.y), z in (lo.z, hi.z))
    idx_corners = [Bmat \ collect(Tuple(c)) for c in corners]
    mins = reduce((a, b) -> min.(a, b), idx_corners)
    maxs = reduce((a, b) -> max.(a, b), idx_corners)
    pad = ceil(Int, lp.r / lp.a) + 1
    lo_idx = floor.(Int, mins) .- pad
    hi_idx = ceil.(Int, maxs) .+ pad
    return (Tuple(lo_idx), Tuple(hi_idx))
end

# --- AABB overlap graph (docs/algorithm.md §6.3, §6.5) ---------------------
#
# Pure Julia, gmsh-free primitives shared by `balanced_fuse!`'s pre-filter
# (pipeline.jl), the boundary-COMMON operand-disjointness split
# (`trim_disjoint`, pipeline.jl), and the floating-body cleanup gate
# (`filter_floating!`, pipeline.jl). Kept here rather than in geomkernel.jl
# so they can be unit tested with no gmsh session at all.

"""
    aabbs_overlap(b1, b2; eps=1e-6) -> Bool

Do two axis-aligned bounding boxes, each `(xmin,ymin,zmin,xmax,ymax,zmax)`,
overlap (touching counts as overlapping, within `eps`)? No OCCT boolean math
— a cheap pre-filter over raw box coordinates."""
function aabbs_overlap(b1::NTuple{6,Float64}, b2::NTuple{6,Float64}; eps::Real=1e-6)
    return b1[1] <= b2[4] + eps && b1[4] >= b2[1] - eps &&
           b1[2] <= b2[5] + eps && b1[5] >= b2[2] - eps &&
           b1[3] <= b2[6] + eps && b1[6] >= b2[3] - eps
end

"""
    overlap_components(tags, boxes; eps=1e-6) -> Vector{Vector{Int32}}

Partition `tags` into connected components of the AABB-overlap graph (two
tags are in the same component iff there is a chain of pairwise
`aabbs_overlap` box overlaps connecting them), via union-find. `boxes` must
contain an entry for every tag in `tags` (see `bounding_boxes`).

This is the shared primitive behind the operand-disjointness invariant
(docs/algorithm.md §6.3): a singleton component is *provably* disjoint from
every other tag, so it is always safe to pass to a boolean call alone or
batched with other singletons; components with >1 member need to be
resolved (fused, or handled one at a time) before touching a boolean kernel
call, since OCCT's multi-operand booleans fragment (rather than union)
mutually-overlapping operands.

Naive O(n²) pairwise comparison, which is fine below a few thousand tags —
this is the common case (one tile or one assembly batch). Callers with
larger `tags` should pre-bucket by AABB center into a uniform grid sized at
the median box diagonal and only run this within/between neighboring
buckets, but no call site in this codebase currently needs that."""
function overlap_components(tags::Vector{Int32}, boxes::Dict{Int32,NTuple{6,Float64}}; eps::Real=1e-6)
    n = length(tags)
    parent = Dict{Int32,Int32}(t => t for t in tags)
    function find(x::Int32)
        while parent[x] != x
            parent[x] = parent[parent[x]]   # path halving
            x = parent[x]
        end
        return x
    end
    function union!(a::Int32, b::Int32)
        ra, rb = find(a), find(b)
        ra != rb && (parent[ra] = rb)
        return nothing
    end
    for i in 1:n, j in i+1:n
        ti, tj = tags[i], tags[j]
        aabbs_overlap(boxes[ti], boxes[tj]; eps=eps) && union!(ti, tj)
    end
    groups = Dict{Int32,Vector{Int32}}()
    for t in tags
        push!(get!(() -> Int32[], groups, find(t)), t)
    end
    return collect(values(groups))
end

"""
    candidate_struts(lp, lo, hi) -> Vector{StrutRef}

Every strut whose origin node index falls within the padded index range
covering the world-space AABB `[lo, hi]` (docs/algorithm.md §2.4). This may
include struts that are trivially outside the true volume — classification
(`classify.jl`) discards those cheaply.
"""
function candidate_struts(lp::LatticeParams, lo::Vec3, hi::Vec3)
    lo_idx, hi_idx = index_range(lp, lo, hi)
    n = (hi_idx[1] - lo_idx[1] + 1) * (hi_idx[2] - lo_idx[2] + 1) * (hi_idx[3] - lo_idx[3] + 1) * 3
    struts = Vector{StrutRef}(undef, 0)
    sizehint!(struts, max(n, 0))
    for i in lo_idx[1]:hi_idx[1], j in lo_idx[2]:hi_idx[2], k in lo_idx[3]:hi_idx[3]
        for d in 1:3
            push!(struts, StrutRef(i, j, k, d))
        end
    end
    return struts
end

File: src/latticegen2/lattice.py
"""Lattice mathematics.

Implements docs/algorithm.md §2 and §3.1 exactly — strut directions, the node
basis, the diamond profile frame, and candidate enumeration from a world-space
AABB. Every angle and length comes from an expression (``asin(sqrt(2/3))``),
never a decimal literal, so precision is IEEE-754 double throughout.

This module is pure NumPy: it never touches the geometry kernel, which is what
makes the lattice math unit-testable on its own (``test/test_lattice.py``).

The one concept this module adds beyond docs/algorithm.md is the **half-strut**:
the half of a strut adjacent to one node, obtained by cutting every strut at its
midpoint. Each node owns exactly six of them (``±e_k`` for ``k = 0, 1, 2``), and
their union is the junction solid ``J`` that the whole fuse-free architecture is
built on — see docs/algorithm.md §3.2.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

# --- Strut recline angle (docs/algorithm.md §2.1) --------------------------

THETA = float(np.arcsin(np.sqrt(2.0 / 3.0)))
"""Exact strut recline angle from +Z, ~54.7356°, computed natively."""

SIN_THETA = float(np.sqrt(2.0 / 3.0))
COS_THETA = float(1.0 / np.sqrt(3.0))


def strut_direction(k: int) -> np.ndarray:
    """Unit direction of strut family ``k in {0,1,2}`` (docs/algorithm.md §2.1).

    Azimuth ``2*pi*k/3`` around Z, reclined from +Z by :data:`THETA`.
    """
    phi = 2.0 * np.pi * k / 3.0
    return np.array(
        [SIN_THETA * np.cos(phi), SIN_THETA * np.sin(phi), COS_THETA], dtype=float
    )


def cube_edge(cc: float) -> float:
    """Cube edge length ``a = cc / sqrt(2)`` (docs/algorithm.md §2.2)."""
    return float(cc) / np.sqrt(2.0)


# --- Precomputed lattice geometry ------------------------------------------


@dataclass(frozen=True)
class LatticeParams:
    """All lattice geometry derived from the CLI parameters ``cc`` and ``t``.

    Attributes are indexed ``0:3`` for strut families ``e0, e1, e2``, matching
    the ``k`` of docs/algorithm.md.

    ``e``, ``u`` and ``v`` are ``(3, 3)`` arrays whose **rows** are the
    per-direction vectors; ``B`` is the ``(3, 3)`` basis matrix whose
    **columns** are ``a * e_k``, so ``node(i,j,k) = B @ [i,j,k]``.
    """

    cc: float
    t: float
    a: float
    """Cube edge length, ``cc / sqrt(2)``."""
    r: float
    """Strut circumradius, ``t / sqrt(2)`` — half the profile's diagonal."""
    e: np.ndarray
    u: np.ndarray
    """Horizontal profile axis per direction (docs/algorithm.md §3.1)."""
    v: np.ndarray
    """Vertical-plane profile axis per direction (docs/algorithm.md §3.1)."""
    B: np.ndarray


def lattice_params(cc: float, t: float) -> LatticeParams:
    """Build a :class:`LatticeParams` from the two CLI geometry parameters."""
    a = cube_edge(cc)
    e = np.array([strut_direction(k) for k in range(3)], dtype=float)
    zhat = np.array([0.0, 0.0, 1.0])
    u = np.array([_normalize(np.cross(zhat, e[k])) for k in range(3)], dtype=float)
    v = np.array([np.cross(e[k], u[k]) for k in range(3)], dtype=float)
    B = a * np.column_stack([e[0], e[1], e[2]])
    return LatticeParams(cc=float(cc), t=float(t), a=a, r=float(t) / np.sqrt(2.0),
                         e=e, u=u, v=v, B=B)


def _normalize(x: np.ndarray) -> np.ndarray:
    return x / np.linalg.norm(x)


def node(lp: LatticeParams, ijk) -> np.ndarray:
    """World-space position of lattice node ``(i,j,k)`` (docs/algorithm.md §2.2)."""
    return lp.B @ np.asarray(ijk, dtype=float)


def nodes(lp: LatticeParams, ijk: np.ndarray) -> np.ndarray:
    """Vectorized :func:`node`: ``(N, 3)`` integer indices -> ``(N, 3)`` positions."""
    return np.asarray(ijk, dtype=float) @ lp.B.T


def profile_vertices(lp: LatticeParams, center: np.ndarray, k: int) -> np.ndarray:
    """The four corners of the diamond cross-section centred at ``center``.

    Order is ``+u, +v, -u, -v`` (docs/algorithm.md §3.1), which traverses the
    square consistently — consecutive entries are adjacent corners, so the list
    is directly usable as a closed polygon.
    """
    hu = lp.r * lp.u[k]
    hv = lp.r * lp.v[k]
    c = np.asarray(center, dtype=float)
    return np.array([c + hu, c + hv, c - hu, c - hv], dtype=float)


# --- Half-struts ------------------------------------------------------------

HALF_STRUTS: tuple[tuple[int, int], ...] = (
    (0, +1), (1, +1), (2, +1), (0, -1), (1, -1), (2, -1),
)
"""The six ``(k, sign)`` half-struts every node owns, in canonical order.

Index ``h`` in this tuple is the half-strut's id throughout the pipeline;
``0..2`` are the outgoing (``+e_k``) halves and ``3..5`` the incoming ones, so
``h`` and ``(h + 3) % 6`` are always the two halves pointing opposite ways along
the same axis.
"""


def half_strut_offset(lp: LatticeParams, h: int) -> np.ndarray:
    """Vector from a node to the far (cap) end of half-strut ``h``."""
    k, s = HALF_STRUTS[h]
    return s * (lp.a / 2.0) * lp.e[k]


def neighbor_step(h: int) -> np.ndarray:
    """Index-space step to the node sharing half-strut ``h``'s cap.

    Half-strut ``h`` of node ``n`` and half-strut ``(h + 3) % 6`` of node
    ``n + neighbor_step(h)`` are the two halves of one strut, and their cap
    quads are the same square in world space.
    """
    k, s = HALF_STRUTS[h]
    step = np.zeros(3, dtype=np.int64)
    step[k] = s
    return step


OPPOSITE_HALF: tuple[int, ...] = tuple((h + 3) % 6 for h in range(6))
"""``OPPOSITE_HALF[h]`` is the half-strut facing ``h`` across the shared cap."""


# --- Candidate index range (docs/algorithm.md §2.4) -------------------------


def index_range(lp: LatticeParams, lo: np.ndarray, hi: np.ndarray):
    """Inclusive ``(lo_idx, hi_idx)`` node-index bounds covering AABB ``[lo, hi]``.

    Ported verbatim from docs/algorithm.md §2.4: map the box's eight corners
    into index space through ``B``, take the floor/ceil hull, and pad by
    ``ceil(r/a) + 1`` cells so no strut that could geometrically touch the
    (circumradius-expanded) volume is missed.

    The padding is comfortably more than the ``a/2`` reach of a junction, so a
    node outside the returned range cannot contribute material inside the box.
    """
    lo = np.asarray(lo, dtype=float)
    hi = np.asarray(hi, dtype=float)
    corners = np.array(
        [[x, y, z] for x in (lo[0], hi[0]) for y in (lo[1], hi[1]) for z in (lo[2], hi[2])],
        dtype=float,
    )
    idx_corners = np.linalg.solve(lp.B, corners.T).T
    pad = int(np.ceil(lp.r / lp.a)) + 1
    lo_idx = np.floor(idx_corners.min(axis=0)).astype(np.int64) - pad
    hi_idx = np.ceil(idx_corners.max(axis=0)).astype(np.int64) + pad
    return lo_idx, hi_idx


def candidate_nodes(lp: LatticeParams, lo: np.ndarray, hi: np.ndarray) -> np.ndarray:
    """Every node index in :func:`index_range`, as an ``(N, 3)`` int array.

    Rows are ordered so the last index varies fastest, which keeps nodes that
    are spatial neighbours close together in memory for the classification
    sweep.
    """
    lo_idx, hi_idx = index_range(lp, lo, hi)
    ii = np.arange(lo_idx[0], hi_idx[0] + 1, dtype=np.int64)
    jj = np.arange(lo_idx[1], hi_idx[1] + 1, dtype=np.int64)
    kk = np.arange(lo_idx[2], hi_idx[2] + 1, dtype=np.int64)
    grid = np.meshgrid(ii, jj, kk, indexing="ij")
    return np.stack([g.ravel() for g in grid], axis=1)


def format_param(x: float) -> str:
    """Format a float for file names and part names without trailing zeros.

    ``20.0 -> "20"``, ``1.5 -> "1.5"`` (docs/algorithm.md §9's naming rules).
    """
    if x == int(x):
        return str(int(x))
    return repr(float(x)).rstrip("0").rstrip(".")


def part_name(input_path: str, cc: float, t: float) -> str:
    """STEP part name ``<input_stem>+cc<cc>+t<t>`` (specification.md §5)."""
    import os

    stem = os.path.splitext(os.path.basename(input_path))[0]
    return f"{stem}+cc{format_param(cc)}+t{format_param(t)}"
